kernel/src/syscalls.c

```diff
@@ -0,0 +1,87 @@
+// A function in the syscall table can have up to 3 64-bit parameters, and return both 64 and 128
+// bits. If you can make sure the ECX parameter is unused, all but the stack parameters can be used.
+
+#include "syscalls.h"
+#include "gdt.h"
+
+// Number of entries in the syscall table, can be increased when needed
+#define NUM_SYSCALLS 16
+
+void* g_syscall_table[NUM_SYSCALLS];
+
+__attribute__((naked)) void syscall_dispatcher() {
+    asm volatile("push %%rsp\n" // Store userspace stack pointer
+                 "push %%rcx\n" // rcx contains rip before syscall
+                 "push %%r11\n" // r11 contains flags before syscall
+                 "push %%rbp\n"
+                 "mov %%rsp, %%rbp\n"
+
+                 // Make sure syscall is within bounds
+                 "cmp %[max_syscalls], %%rax\n"
+                 "jg _oob\n"
```

Comment:
This should be greater or equal because a value equal to NUM_SYSCALLS is not of range of the table.